pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,564 @@
+		initialized: false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
+		return nil
+	}
+
+	err := m.storage.Init(context.Background(), "cs3-share-manager-metadata")
+	if err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "shares"); err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "metadata"); err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "OwnerId",
+		Func: indexOwnerFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "GranteeId",
+		Func: indexGranteeFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	m.initialized = true
+	return nil
+}
+
+// Share creates a new share
+func (m *Manager) Share(ctx context.Context, md *provider.ResourceInfo, g *collaboration.ShareGrant) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+	user := ctxpkg.ContextMustGetUser(ctx)
+	now := time.Now().UnixNano()
+	ts := &typespb.Timestamp{
+		Seconds: uint64(now / 1000000000),
+		Nanos:   uint32(now % 1000000000),
+	}
+
+	share := &collaboration.Share{
+		Id: &collaboration.ShareId{
+			OpaqueId: uuid.NewString(),
+		},
+		ResourceId:  md.Id,
+		Permissions: g.Permissions,
+		Grantee:     g.Grantee,
+		Owner:       md.Owner,
+		Creator:     user.Id,
+		Ctime:       ts,
+		Mtime:       ts,
+	}
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
+	err = m.storage.SimpleUpload(ctx, fn, data)
+	if err != nil {
+		return nil, err
+	}
+
+	metadataPath := path.Join("metadata", share.Id.OpaqueId)
+	err = m.storage.MakeDirIfNotExist(ctx, metadataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = m.indexer.Add(share)
+	if err != nil {
+		return nil, err
+	}
+
+	return share, nil
+}
+
+// GetShare gets the information for a share by the given ref.
+func (m *Manager) GetShare(ctx context.Context, ref *collaboration.ShareReference) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+
+	switch {
+	case ref.GetId() != nil:
+		return m.getShareByID(ctx, ref.GetId().OpaqueId)
+	case ref.GetKey() != nil:
+		return m.getShareByKey(ctx, ref.GetKey())
+	default:
+		return nil, errtypes.BadRequest("neither share id nor key was given")
+	}
+}
+
+// Unshare deletes the share pointed by ref.
+func (m *Manager) Unshare(ctx context.Context, ref *collaboration.ShareReference) error {
+	if err := m.initialize(); err != nil {
+		return err
+	}
+	share, err := m.GetShare(ctx, ref)
+	if err != nil {
+		return err
+	}
+
+	fn := path.Join("shares", ref.GetId().OpaqueId)
+	err = m.storage.Delete(ctx, fn)
+	if err != nil {
+		return err
+	}
+
+	return m.indexer.Delete(share)
```

Comment:
Are we sure we want to error if the indexer fails to delete but the storage was successful? What should the client do with that information?